### PR TITLE
Automatically annotate Services for NPL

### PR DESCRIPTION
**Pull request by monotosh-avi**
AKO would automatically annotate required services for NPL. These
are Services of type LB and backends for ingresses.

- For Service type LB in check if the service is annotated, else
annotate and return.
- For ingress add, annotate all backend services and return.
- For ingress deletion check if annotation can be removed from bakcend service. This
is done after checking if the service is referred by any other ingress.
- For ingress update, check if any new service requires annotation, and if
annotation from old service can be be removed.

Helm template and operator changes to disable this feature by setting autoAnnotateService to false
would be added in next PR

### CODE REVIEW

**Comment by monotosh-avi** [comment]
build ako

**Comment by monotosh-avi** [comment]
build ako

**Comment by monotosh-avi** [comment]
build ako

**Comment by monotosh-avi** [comment]
build ako

**Comment by monotosh-avi** [comment]
build ako

**Comment by monotosh-avi** [comment]
build ako

**Comment by monotosh-avi** [comment]
build ako

File: internal/status/svc_annotation.go [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+	patchPayload := map[string]interface{}{
+		"metadata": map[string]map[string]string{
+			"annotations": ann,
+		},
+	}
+
+	payloadBytes, _ := json.Marshal(patchPayload)
+	_, err = utils.GetInformers().ClientSet.CoreV1().Services(service.Namespace).Patch(context.TODO(), service.Name, types.MergePatchType, payloadBytes, metav1.PatchOptions{})
+	if err != nil {
+		utils.AviLog.Errorf("key: %s, msg: there was an error in updating the Service annotation for NPL: %v", key, err)
```
Comment (sudswasavi) [inline internal/status/svc_annotation.go:54]: you should probably return here? Else the log below will get printed.

File: internal/status/svc_annotation.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
```
Comment (sudswasavi) [inline internal/status/svc_annotation.go:82]: Same here, if it's errors out - we should return.

File: internal/status/svc_annotation.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+	}
+	ann[lib.NPLSvcAnnotation] = "true"
+	patchPayload := map[string]interface{}{
+		"metadata": map[string]map[string]string{
+			"annotations": ann,
+		},
+	}
+
+	payloadBytes, _ := json.Marshal(patchPayload)
+	_, err = utils.GetInformers().ClientSet.CoreV1().Services(service.Namespace).Patch(context.TODO(), service.Name, types.MergePatchType, payloadBytes, metav1.PatchOptions{})
```
Comment (sudswasavi) [inline internal/status/svc_annotation.go:52]: Can listers get used instead of the clientset?

File: internal/status/svc_annotation.go (hunk 2 above)
Comment (monotosh-avi) [reply internal/status/svc_annotation.go:82]: Added

File: internal/status/svc_annotation.go (hunk 3 above)
Comment (monotosh-avi) [reply internal/status/svc_annotation.go:52]: I don't think we can use lister, that can be used to get services only.

File: internal/status/svc_annotation.go (hunk 1 above)
Comment (monotosh-avi) [reply internal/status/svc_annotation.go:54]: Done

**Review by sudswasavi — APPROVED** [review]

File: tests/npltests/npl_pod_test.go [hunk 4]
```diff
@@ -12,7 +12,7 @@
 * limitations under the License.
 */
 
-package hostnameshardtests
+package npltests
```
Comment (shubhamavi) [inline tests/npltests/npl_pod_test.go:15]: could you please rebase once, i remember doing this change and some addition regarding CRDClientset in a previous PR

File: internal/status/svc_annotation.go [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+	if err != nil {
+		utils.AviLog.Errorf("key: %s, msg: there was an error in updating the Service annotation for NPL: %v", key, err)
+		return
+	}
+	utils.AviLog.Debugf("key: %s, msg: Deleted NPL annotation from Service: %s/%s", namespace, name)
```
Comment (shubhamavi) [inline internal/status/svc_annotation.go:86]: key is missing in the formatting argument.

File: internal/status/svc_annotation.go [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+	service, err := utils.GetInformers().ServiceInformer.Lister().Services(namespace).Get(name)
+	if err != nil {
+		return
+	}
+	ann := service.GetAnnotations()
+	if _, found := ann[lib.NPLSvcAnnotation]; !found {
+		return
+	}
+	if ann == nil {
+		ann = make(map[string]string)
```
Comment (shubhamavi) [inline internal/status/svc_annotation.go:71]: can we return from here?

File: internal/nodes/ingress_model_rel.go [hunk 7]
```diff
@@ -307,10 +313,28 @@ func IngressChanges(ingName string, namespace string, key string) ([]string, boo
...
-			utils.AviLog.Debugf("key: %s, msg: updating ingress relationship for service: %s", key, svc)
+		_, oldSvcs := objects.SharedSvcLister().IngressMappings(namespace).GetIngToSvc(ingName)
+		currSvcs := parseServicesForIngress(ingObj.Spec, key)
+
+		svcToDel := lib.Difference(oldSvcs, currSvcs)
+		for _, svc := range svcToDel {
+			_, ingrforSvc := objects.SharedSvcLister().IngressMappings(namespace).GetSvcToIng(svc)
+			ingrforSvc = utils.Remove(ingrforSvc, ingName)
+			//objects.SharedSvcLister().IngressMappings(namespace).UpdateSvcToIngMapping(svc, ingrforSvc)
+			if len(ingrforSvc) == 0 {
```
Comment (shubhamavi) [inline internal/nodes/ingress_model_rel.go:324]: should we be checking for AutoAnnotateNPLSvc() here as well? also let's remove the commented out code above.

File: internal/nodes/ingress_model_rel.go [hunk 8]
```diff
@@ -291,7 +292,12 @@ func IngressChanges(ingName string, namespace string, key string) ([]string, boo
 		if k8serrors.IsNotFound(err) {
 			// Remove all the Ingress to Services mapping.
 			// Remove the references of this ingress from the Services
-			objects.SharedSvcLister().IngressMappings(namespace).RemoveIngressMappings(ingName)
+			svcToDel := objects.SharedSvcLister().IngressMappings(namespace).RemoveIngressMappings(ingName)
```
Comment (shubhamavi) [inline internal/nodes/ingress_model_rel.go:295]: this can move inside the conditional below

File: internal/lib/lib.go [hunk 9]
```diff
@@ -550,10 +550,20 @@ func IsNodePortMode() bool {
...
+// ToDo: Set the Service Type only once. But this creates a problem in UTs,
+// because different types of Services needs to be tested in the UTs.
 func GetServiceType() string {
 	return os.Getenv(SERVICE_TYPE)
 }
 
+func AutoAnnotateNPLSvc() bool {
+	autoAnnotateSvc := os.Getenv(autoAnnotateService)
+	if GetServiceType() == NodePortLocal && autoAnnotateSvc != "false" {
+		return true
```
Comment (shubhamavi) [inline internal/lib/lib.go:562]: this will return true if the user passes the value as "False".

Comment (monotosh-avi) [reply internal/lib/lib.go:562]: corrected to do case insensitive match

File: internal/nodes/ingress_model_rel.go (hunk 8 above)
Comment (monotosh-avi) [reply internal/nodes/ingress_model_rel.go:295]: We need to remove the mapping even if NPL is not enabled, so the function call has to be made, the return value can be ignored.

File: internal/nodes/ingress_model_rel.go (hunk 7 above)
Comment (monotosh-avi) [reply internal/nodes/ingress_model_rel.go:324]: Good catch ! Added the condition.

File: internal/status/svc_annotation.go (hunk 6 above)
Comment (monotosh-avi) [reply internal/status/svc_annotation.go:71]: Yeah, we can. Added.

File: internal/status/svc_annotation.go (hunk 5 above)
Comment (monotosh-avi) [reply internal/status/svc_annotation.go:86]: Added

File: tests/npltests/npl_pod_test.go (hunk 4 above)
Comment (monotosh-avi) [reply tests/npltests/npl_pod_test.go:15]: I think that PR is not merged yet :-)

**Review by sudswasavi — APPROVED** [review]

File: tests/npltests/npl_pod_test.go (hunk 4 above)
Comment (shubhamavi) [reply tests/npltests/npl_pod_test.go:15]: ah, i'll rebase then :) thanks

**Review by shubhamavi — APPROVED** [review]

**Pull request merged**